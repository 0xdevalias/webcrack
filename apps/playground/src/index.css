@tailwind base;
@tailwind components;
@tailwind utilities;

html, body, #root {
  height: 100%;
}

#root {
  height: 100vh;
  width: 100vw;
  display: flex;
  flex-direction: column;
}
