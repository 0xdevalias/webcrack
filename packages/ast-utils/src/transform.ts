import traverse, { Node, TraverseOptions, visitors } from '@babel/traverse';

export async function applyTransformAsync<TOptions>(
  ast: Node,
  transform: AsyncTransform<TOptions>,
  options?: TOptions,
): Promise<TransformState> {
  const state: TransformState = { changes: 0 };

  await transform.run?.(ast, state, options);
  if (transform.visitor)
    traverse(ast, transform.visitor(options), undefined, state);

  return state;
}

export function applyTransform<TOptions>(
  ast: Node,
  transform: Transform<TOptions>,
  options?: TOptions,
): TransformState {
  const state: TransformState = { changes: 0 };

  transform.run?.(ast, state, options);

  if (transform.visitor) {
    const visitor = transform.visitor(options);
    visitor.noScope = !transform.scope;
    traverse(ast, visitor, undefined, state);
  }

  return state;
}

export function applyTransforms(
  ast: Node,
  transforms: Transform[],
): TransformState {
  const state: TransformState = { changes: 0 };

  for (const transform of transforms) {
    transform.run?.(ast, state);
  }

  const traverseOptions = transforms.flatMap((t) => t.visitor?.() ?? []);
  if (traverseOptions.length > 0) {
    const visitor: TraverseOptions<TransformState> =
      visitors.merge(traverseOptions);
    visitor.noScope = transforms.every((t) => !t.scope);
    traverse(ast, visitor, undefined, state);
  }

  return state;
}

export interface TransformState {
  changes: number;
}

export interface Transform<TOptions = unknown> {
  name: string;
  tags: Tag[];
  scope?: boolean;
  run?: (ast: Node, state: TransformState, options?: TOptions) => void;
  visitor?: (options?: TOptions) => TraverseOptions<TransformState>;
}

export interface AsyncTransform<TOptions = unknown>
  extends Transform<TOptions> {
  run?: (ast: Node, state: TransformState, options?: TOptions) => Promise<void>;
}

export type Tag = 'safe' | 'unsafe';
