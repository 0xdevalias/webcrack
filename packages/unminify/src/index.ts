import { parse } from "@babel/parser";
import {
  applyTransform,
  applyTransforms,
  generate,
  Transform,
} from "@webcrack/ast-utils";
import * as transforms from "./transforms";

export const unminify = {
  name: "unminify",
  tags: ["safe"],
  scope: true,
  run(ast, state) {
    state.changes += applyTransforms(ast, Object.values(transforms)).changes;
  },
} satisfies Transform;

export function unminifySource(code: string): string {
  const ast = parse(code, {
    sourceType: "unambiguous",
    allowReturnOutsideFunction: true,
    plugins: ["jsx"],
  });

  applyTransform(ast, unminify);

  return generate(ast);
}
