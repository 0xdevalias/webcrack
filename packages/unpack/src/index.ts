import { parse } from "@babel/parser";
import traverse, { Visitor, visitors } from "@babel/traverse";
import * as t from "@babel/types";
import * as m from "@codemod/matchers";
import { TransformState } from "@webcrack/ast-utils";
import { unpackBrowserify } from "./browserify";
import { Bundle } from "./bundle";
import { unpackWebpack } from "./webpack";

export { Bundle } from "./bundle";

export function unpack(
  code: string,
  mappings: Record<string, m.Matcher<unknown>> = {},
): Bundle | undefined {
  const ast = parse(code, {
    sourceType: "unambiguous",
    allowReturnOutsideFunction: true,
    plugins: ["jsx"],
  });
  return unpackAST(ast, mappings);
}

export function unpackAST(
  ast: t.Node,
  mappings: Record<string, m.Matcher<unknown>> = {},
): Bundle | undefined {
  const options: { bundle: Bundle | undefined } = { bundle: undefined };
  const traverseOptions: Visitor<TransformState>[] = [
    unpackWebpack.visitor(options),
    unpackBrowserify.visitor(options),
  ];
  const visitor = visitors.merge(traverseOptions);
  traverse(ast, visitor, undefined, { changes: 0 });
  // TODO: applyTransforms(ast, [unpackWebpack, unpackBrowserify]) instead
  if (options.bundle) {
    options.bundle.applyMappings(mappings);
    options.bundle.applyTransforms();
  }
  return options.bundle;
}
