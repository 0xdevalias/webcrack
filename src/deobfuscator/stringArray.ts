import traverse, { NodePath } from '@babel/traverse';
import * as t from '@babel/types';
import * as m from '@codemod/matchers';

export interface StringArray {
  path: NodePath<t.FunctionDeclaration>;
  references: NodePath[];
  name: string;
  strings: string[];
}

export function findStringArray(ast: t.Node) {
  let result: StringArray | undefined;

  traverse(ast, {
    FunctionDeclaration(path) {
      if (matcher.match(path.node)) {
        const strings = array.current!.elements.map(
          e => (e as t.StringLiteral).value
        );
        const name = functionName.current!;
        let references = path.parentPath.scope.bindings[name].referencePaths;
        // Skip references in the getStringArray function itself
        references = references.filter(ref => !ref.findParent(p => p === path));
        path.parentPath.scope.rename(name, '__STRING_ARRAY__');
        result = { path, references, name: '__STRING_ARRAY__', strings };
        path.stop();
      }
    },
  });

  return result;
}

const functionName = m.capture(m.anyString());
const arrayName = m.capture(m.anyString());
const array = m.capture(m.arrayExpression(m.arrayOf(m.stringLiteral())));
// getStringArray = function () { return n; };
const functionAssignment = m.assignmentExpression(
  '=',
  m.identifier(m.fromCapture(functionName)),
  m.functionExpression(
    undefined,
    [],
    m.blockStatement([
      m.returnStatement(m.identifier(m.fromCapture(arrayName))),
    ])
  )
);
const variableDeclaration = m.variableDeclaration(undefined, [
  m.variableDeclarator(m.identifier(arrayName), array),
]);
// function getStringArray() { ... }
const matcher = m.functionDeclaration(
  m.identifier(functionName),
  [],
  m.or(
    // var array = ["hello", "world"];
    // return (getStringArray = function () { return array; })();
    m.blockStatement([
      variableDeclaration,
      m.returnStatement(m.callExpression(functionAssignment)),
    ]),
    // var array = ["hello", "world"];
    // getStringArray = function () { return n; });
    // return getStringArray();
    m.blockStatement([
      variableDeclaration,
      m.expressionStatement(functionAssignment),
      m.returnStatement(m.callExpression(m.identifier(functionName))),
    ])
  )
);
