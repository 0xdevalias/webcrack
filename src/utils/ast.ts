import generate from '@babel/generator';
import * as t from '@babel/types';

export function codePreview(node: t.Node) {
  const { code } = generate(node, { compact: true, comments: false });
  if (code.length > 100) {
    return code.slice(0, 70) + ' â€¦ ' + code.slice(-30);
  }
  return code;
}
