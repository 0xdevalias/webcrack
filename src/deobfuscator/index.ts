import traverse from '@babel/traverse';
import * as t from '@babel/types';
import * as m from '@codemod/matchers';
import extractTernaryCalls from '../transforms/extractTernaryCalls';
import { codePreview } from '../utils/ast';
import { findArrayRotator } from './arrayRotator';
import { findDecoder } from './decoder';
import { findStringArray } from './stringArray';
import { createVM } from './vm';

// https://astexplorer.net/#/gist/b1018df4a8daebfcb1daf9d61fe17557/4ff9ad0e9c40b9616956f17f59a2d9888cd62a4f

export default (ast: t.Node) => {
  const stringArray = findStringArray(ast);
  console.log(`String Array: ${!!stringArray}`);
  if (!stringArray) return;
  console.log(' - length: ' + stringArray.strings.length);
  console.log(` - ${codePreview(stringArray.path.node)}`);

  const rotator = findArrayRotator(ast);
  console.log(`String Array Rotate: ${!!rotator}`);
  if (rotator) {
    console.log(` - ${codePreview(rotator.path.node)}`);
  }

  const decoder = findDecoder(stringArray);
  console.log(`String Array Decode: ${!!decoder}`);
  if (!decoder) return;
  console.log(` - ${codePreview(decoder.path.node)}`);
  decoder.inlineAliasVars();
  console.log('Inlined decoder alias vars');

  const vm = createVM({ stringArray, rotator, decoder });

  traverse(
    ast,
    extractTernaryCalls.visitor(options => options.callee === decoder.name),
    undefined,
    { changes: 0 }
  );

  traverse(ast, {
    CallExpression(path) {
      const matcher = m.callExpression(
        m.identifier(decoder.name),
        m.anything()
      );

      if (matcher.match(path.node) && t.isLiteral(path.node.arguments[0])) {
        const args = path.node.arguments.map(arg => {
          if (t.isNumericLiteral(arg) || t.isStringLiteral(arg)) {
            return arg.value;
          }
        });
        path.replaceWith(t.stringLiteral(vm.decode(args)));
      }
    },
    noScope: true,
  });

  stringArray.path.remove();
  decoder.path.remove();
  rotator?.path.remove();
};
