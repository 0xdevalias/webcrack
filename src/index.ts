import generate from '@babel/generator';
import { parse } from '@babel/parser';
import * as m from '@codemod/matchers';
import { mkdir, writeFile } from 'node:fs/promises';
import { join } from 'node:path';
import deobfuscator from './deobfuscator';
import debugProtection from './deobfuscator/debugProtection';
import selfDefending from './deobfuscator/selfDefending';
import { Bundle, extractBundle } from './extractor';
import { applyTransform, applyTransforms } from './transforms';
import { resetRunState } from './transforms/index';

export interface WebcrackResult {
  code: string;
  bundle: Bundle | undefined;
  /**
   * Save the deobufscated code and the extracted bundle to the given directory.
   * @param path Output directory
   */
  save(path: string): Promise<void>;
}

export interface Options {
  /**
   * Run for every module after generating the code and before saving it.
   * This can be used to format the code or apply other transformations.
   */
  transformCode?: (code: string) => Promise<string> | string;
  /**
   * Assigns paths to modules based on the given matchers.
   * This will also rewrite `require()` calls to use the new paths.
   *
   * @example
   * ```js
   * m => ({
   *   './utils/color.js': m.regExpLiteral('^#([0-9a-f]{3}){1,2}$')
   * })
   * ```
   */
  mappings?: (
    m: typeof import('@codemod/matchers')
  ) => Record<string, m.Matcher<any>>;
  /**
   * Maximum number of iterations for readability transforms.
   */
  maxIterations?: number;
}

export const defaultOptions = {
  maxIterations: 5,
} satisfies Options;

export async function webcrack(
  code: string,
  options: Options = defaultOptions
): Promise<WebcrackResult> {
  const ast = parse(code, {
    sourceType: 'unambiguous',
    allowReturnOutsideFunction: true,
  });

  applyTransform(ast, deobfuscator);

  resetRunState();
  for (
    let i = 1;
    i <= (options.maxIterations ?? defaultOptions.maxIterations);
    i++
  ) {
    console.log('\n== Iteration', i, '==');
    if (applyTransforms(ast, ['readability']).changes === 0) break;
  }

  // Have to run this after readability transforms because the function may contain dead code
  applyTransform(ast, selfDefending);
  applyTransform(ast, debugProtection);

  resetRunState();

  const bundle = extractBundle(ast);
  console.log('Bundle:', bundle?.type);

  let outputCode = generate(ast).code;
  outputCode = options.transformCode
    ? await options.transformCode(outputCode)
    : outputCode;

  return {
    code: outputCode,
    bundle,
    async save(path) {
      await mkdir(path, { recursive: true });
      await writeFile(join(path, 'deobfuscated.js'), outputCode, 'utf8');
      bundle?.save(path, options.transformCode, options.mappings);
    },
  };
}
