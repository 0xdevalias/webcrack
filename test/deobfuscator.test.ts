import generate from '@babel/generator';
import { parse } from '@babel/parser';
import traverse from '@babel/traverse';
import { readFile } from 'fs/promises';
import { describe, expect, test } from 'vitest';
import { findStringArray } from '../src/deobfuscator/stringArray';
import { webcrack } from '../src/index';
import { inlineFunctionAliases, inlineVariableAliases } from '../src/utils/ast';

describe('find string array', async () => {
  const ast = parse(await readFile('./test/samples/obfuscator.io.js', 'utf8'));

  test('function wrapper', () => {
    const stringArray = findStringArray(ast);
    expect(stringArray).toBeDefined();
    expect(stringArray!.name).toBe('__STRING_ARRAY__');
    expect(stringArray!.references).toHaveLength(2);
    expect(stringArray!.strings).toHaveLength(25);
  });
});

describe('encoders', () => {
  test('multiple', async () => {
    const result = webcrack(
      await readFile('./test/samples/obfuscator.io-multi-encoders.js', 'utf8')
    );
    expect(result.code).toMatchSnapshot();
  });
});

describe('inline decoder', () => {
  test('inline variable', () => {
    const ast = parse(`
      function decoder() {}
      decoder(1);
      (() => {
        const alias = decoder;
        alias(2);
        (() => {
          const alias2 = alias;
          alias2(3);
        });
      });
  `);
    traverse(ast, {
      FunctionDeclaration(path) {
        const binding = path.scope.parent.bindings.decoder;
        inlineVariableAliases(binding);
        path.stop();
      },
    });
    expect(generate(ast).code).toMatchSnapshot();
  });

  test('inline function', () => {
    const ast = parse(`
      function decoder() {}
      decoder(1);
      (() => {
        function alias(a, b) {
          return decoder(a - 625, b);
        }
        alias(2, 3);
        (() => {
          function alias2(a, b) {
            return alias(b - -678, a);
          }
          alias2(4, 5);
        })();
      })();
  `);
    traverse(ast, {
      FunctionDeclaration(path) {
        const binding = path.scope.parent.bindings.decoder;
        inlineFunctionAliases(binding);
        path.stop();
      },
    });

    expect(generate(ast).code).toMatchSnapshot();
  });
});
