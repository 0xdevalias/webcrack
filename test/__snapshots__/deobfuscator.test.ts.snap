// Vitest Snapshot v1

exports[`deobfuscate obfuscator.io.js 1`] = `
"// v4.0.0

function hi() {
  console.log(\\"Hello World!\\");
  console.log( /*webcrack:decode_error*/undefined);
  console.log( /*webcrack:decode_error*/undefined);
}
hi();
function a(b) {
  function e(f) {
    if (typeof f === \\"string\\") {
      return function (g) {}.constructor(\\"while (true) {}\\").apply(\\"counter\\");
    } else if ((\\"\\" + f / f).length !== 1 || f % 20 === 0) {
      (function () {
        return true;
      }).constructor(\\"debugger\\").call(\\"action\\");
    } else {
      (function () {
        return false;
      }).constructor(\\"debugger\\").apply(\\"stateObject\\");
    }
    e(++f);
  }
  try {
    if (b) {
      return e;
    } else {
      e(0);
    }
  } catch (f) {}
}"
`;

exports[`deobfuscate obfuscator.io-control-flow.js 1`] = `
"function applyTransforms() {
  this.modules.forEach(varInjection_1.inlineVarInjections);
  this.modules.forEach(esm_1.convertESM);
  (0, getDefaultExport_1.convertDefaultRequire)(this);
  this.replaceRequireCalls();
  var f = m.capture(m.numericLiteral());
  var g = m.callExpression(m.identifier(\\"require\\"), [f]);
  return g;
}"
`;

exports[`deobfuscate obfuscator.io-function-wrapper.js 1`] = `
"function hi() {
  console.log(\\"Hello World!\\");
  console.log(30);
  console.log(undefined);
}
hi();"
`;

exports[`deobfuscate obfuscator.io-high.js 1`] = `
"(function () {
  var _0x5b07a9 = function () {
    var _0x3d6989;
    try {
      _0x3d6989 = Function(\\"return (function() {}.constructor(\\\\\\"return this\\\\\\")( ));\\")();
    } catch (_0x4516e9) {
      _0x3d6989 = window;
    }
    return _0x3d6989;
  };
  var _0x152a02 = _0x5b07a9();
  _0x152a02.setInterval(_0x40ea98, 4000);
})();
function hi() {
  console.log(\\"Hello World!\\");
}
hi();
function _0x40ea98(_0x3857c8) {
  function _0xb25c5e(_0x22f5f3) {
    if (typeof _0x22f5f3 === \\"string\\") {
      return function (_0x40f974) {}.constructor(\\"while (true) {}\\").apply(\\"counter\\");
    } else if ((\\"\\" + _0x22f5f3 / _0x22f5f3).length !== 1 || _0x22f5f3 % 20 === 0) {
      (function () {
        return true;
      }).constructor(\\"debugger\\").call(\\"action\\");
    } else {
      (function () {
        return false;
      }).constructor(\\"debugger\\").apply(\\"stateObject\\");
    }
    _0xb25c5e(++_0x22f5f3);
  }
  try {
    if (_0x3857c8) {
      return _0xb25c5e;
    } else {
      _0xb25c5e(0);
    }
  } catch (_0x3e75bc) {}
}"
`;

exports[`deobfuscate obfuscator.io-multi-encoders.js 1`] = `
"function hi() {
  console.log(\\"Hello World!\\");
}
hi();"
`;

exports[`deobfuscate obfuscator.io-rotator-unary.js 1`] = `
"function hi() {
  console.log(\\"Hello World!\\");
}
hi();"
`;

exports[`inline decoder > inline function 1`] = `
"function decoder() {}
decoder(1);
(() => {
  decoder(2 - 625, 3);
  (() => {
    decoder(5 - -678 - 625, 4);
  })();
})();"
`;

exports[`inline decoder > inline variable 1`] = `
"function decoder() {}
decoder(1);
() => {
  decoder(2);
  decoder(3);
  () => {
    alias(4);
  };
};"
`;
